eng/custom-checkstyles/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientChecks.java

```diff
@@ -44,117 +46,119 @@ public ServiceClientChecks() throws ClassNotFoundException {
-        if (classNameNode == null) {
-            classNameNode = CustomCheckUtils.findPreviousSiblingOfType(classIdentifierNode, node -> node.getType() == TokenTypes.IDENT);
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.PACKAGE_DEF:
+                this.packageName = FullIdent.createFullIdent(token.findFirstToken(TokenTypes.DOT)).getText();
+                break;
+            case TokenTypes.CLASS_DEF:
+                if (extendsServiceClient(token)) {
+                    this.extendsServiceClient = true;
+                }
+                break;
+            case TokenTypes.CTOR_DEF:
+                if (this.extendsServiceClient && visibilityIsPublicOrProtected(token)) {
+                    log(token, CONSTRUCTOR_ERROR_MESSAGE);
+                }
+                break;
+            case TokenTypes.METHOD_DEF:
+                if (this.extendsServiceClient && !this.hasStaticBuilder && methodIsStaticBuilder(token)) {
+                    this.hasStaticBuilder = true;
+                }
+                break;
         }
+    }
 
-        // If the class name cannot be found we shouldn't attempt to run the check.
-        if (classNameNode == null) {
-            return Optional.empty();
+    /**
+     * End of the TreeWalker traversal.
+     * @param rootAST Root of the AST.
+     */
+    @Override
+    public void finishTree(DetailAST rootAST) {
+        if (this.extendsServiceClient && !this.hasStaticBuilder) {
+            log(0, BUILDER_ERROR_MESSAGE);
         }
-
-        return Optional.of(packageName + "." + classNameNode.getText());
     }
 
     /**
-     * Checks if the constructor node has public or protected modifiers.
-     * @param constructorNode Constructor node
+     * Determines if the class extends ServiceClient.
+     * @param classDefinitionToken Class definition token.
+     * @return True if the class is not an inner class, is not in a track one package, and extends ServiceClient.
      */
-    private void checkConstructorIsHidden(DetailAST constructorNode) {
-        DetailAST modifierNode = constructorNode.findFirstToken(TokenTypes.MODIFIERS);
-        if (modifierNode == null) {
-            return;
+    private boolean extendsServiceClient(DetailAST classDefinitionToken) {
+        // Prevent inner classes from stating its parent class extends ServiceClient or if it is in a track one package.
+        if (ScopeUtil.isInClassBlock(classDefinitionToken) || this.packageName.startsWith("com.microsoft")) {
+            return false;
         }
 
-        Optional<DetailAST> disallowedModifierNode = TokenUtil.findFirstTokenByPredicate(modifierNode, node -> {
-            return node.getType() == TokenTypes.LITERAL_PUBLIC || node.getType() == TokenTypes.LITERAL_PROTECTED;
-        });
+        try {
+            String className = this.packageName + "." + classDefinitionToken.findFirstToken(TokenTypes.IDENT).getText();
```

Comment:
Is there no better way of getting the fully-qualified class name than this?

Reply:
Do you mean the concatenation or how I am getting the class name? If it is the latter the answer is no, looking at how checkstyle implements their [AbstractClassNameCheck](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbstractClassNameCheck.java#L154) it is doing the same thing.